### PR TITLE
[coqdep] Refactor initialization API

### PR DESCRIPTION
**Pull request by ejgallego**
We move adding the .v files to its own step; and some cleanups in
the way to the way coqdep init process is handled.

We also provide an API to get the dependencies of the files as an OCaml record without having to print them.

This is a step towards being able to use `coqdep` as a library.

Co-authored-by: Ali Caglayan

### CODE REVIEW

**Comment by ejgallego** [comment]
cc #13364

**Comment by ejgallego** [comment]
@SkySkimmer maybe you could assign this? If not, who would be the person?

Actually maybe even @Alizter could take of it, but not sure how we do the rules on that these days.

File: tools/coqdep/common.mli [hunk 1]
```diff
@@ -50,5 +49,29 @@ val add_rec_dir_import :
...
 val sort : unit -> unit
 
-(** Init coqdep, including argument parsing *)
-val init : unit -> unit
+(** [init args] Init coqdep, parsing arguments from [args]; returns
+   the list of .v files passed as arguments *)
+val init : string list -> string list
+
+(*  *)
+val usage : unit -> unit
```
Comment (SkySkimmer) [inline tools/coqdep/common.mli:57]: empty comment

File: tools/coqdep/common.mli [hunk 2]
```diff
@@ -50,5 +49,29 @@ val add_rec_dir_import :
...
+  | Other of string   (* filenames of dependencies, separated by spaces *)
+end
+
+module Dep_info : sig
+  type t =
+    { name : string
+    ; deps : Dep.t list
+    }
+
+  val print : Format.formatter -> t -> unit
```
Comment (SkySkimmer) [inline tools/coqdep/common.mli:74]:
```suggestion
  (** print in makefile format *)
  val print : Format.formatter -> t -> unit
```

File: tools/coqdep/common.ml [hunk 3]
```diff
@@ -526,21 +528,35 @@ let rec find_dependencies basename =
...
-       printf "%s.vio: %s.v %s\n" ename ename
-         (string_of_dependency_list ".vio" deps);
-       if !write_vos then
-         printf "%s.vos %s.vok %s.required_vos: %s.v %s\n" ename ename ename ename
-           (string_of_dependency_list ".vos" deps);
-       printf "%!")
-    (List.rev !vAccu)
+module Dep_info = struct
+
+  (* This should be put into a Coq_module.t *)
```
Comment (SkySkimmer) [inline tools/coqdep/common.ml:533]: ?

**Comment by Alizter** [comment]
I ran a bench before and nothing has changed (as expected).

<details>
<summary>Details</summary>

```coq
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬───────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │    mem faults     │
│                             │                         │                                       │                                       │                         │                   │
│        package_name         │   NEW      OLD    PDIFF │      NEW             OLD        PDIFF │      NEW             OLD        PDIFF │   NEW      OLD    PDIFF │ NEW  OLD   PDIFF  │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼───────────────────┤
│              coq-coquelicot │   34.50    34.73  -0.66 │   148836807565    148845791575  -0.01 │   196759060298    196780873997  -0.01 │  757700   755900   0.24 │ 109    0      nan │
│                   coq-verdi │   48.91    49.22  -0.63 │   212151509186    213322387595  -0.55 │   319084283682    319137805250  -0.02 │  838692   840792  -0.25 │   0    0      nan │
│      coq-mathcomp-character │   69.34    69.72  -0.55 │   303572746301    305382290372  -0.59 │   464773031636    464754893648   0.00 │  830976   830520   0.05 │   0    0      nan │
│         coq-metacoq-erasure │   63.65    63.99  -0.53 │   273546217268    274766495152  -0.44 │   417975411503    417788052257   0.04 │ 1086700  1086856  -0.01 │   0    0      nan │
│                  coq-stdlib │  416.08   418.06  -0.47 │  1470981308029   1471537197565  -0.04 │  1417136806932   1426942743464  -0.69 │  608340   606460   0.31 │   0    0      nan │
│          coq-mathcomp-field │  102.23   102.67  -0.43 │   447572555316    449591084893  -0.45 │   738949986174    738914654994   0.00 │  707784   708716  -0.13 │   0    0      nan │
│            coq-math-classes │   93.29    93.68  -0.42 │   404342452954    406371461952  -0.50 │   571464003258    571497344229  -0.01 │  502164   503040  -0.17 │   0    0      nan │
│               coq-fiat-core │   57.05    57.28  -0.40 │   236217483564    235512811246   0.30 │   332532190286    332531529793   0.00 │  480360   480224   0.03 │   1    0      nan │
│       coq-engine-bench-lite │  181.76   182.49  -0.40 │   763323044830    767280529642  -0.52 │  1491496988238   1491147542079   0.02 │ 1357712  1357572   0.01 │   0    0      nan │
│                 coq-bignums │   28.60    28.71  -0.38 │   124919881225    125395746186  -0.38 │   171999908326    171999211348   0.00 │  469056   468448   0.13 │   0    0      nan │
│                  coq-geocoq │  718.64   721.08  -0.34 │  3137723691899   3145991251712  -0.26 │  4990997742491   4990939987169   0.00 │ 1076396  1076284   0.01 │   3    0      nan │
│         coq-category-theory │ 1688.05  1692.82  -0.28 │  7395747346134   7412805113177  -0.23 │ 15546575849195  15546415077957   0.00 │ 1468988  1469084  -0.01 │   0    0      nan │
│                coq-coqprime │   43.95    44.06  -0.25 │   190510144920    191073266269  -0.29 │   283332437013    283342291577  -0.00 │  761520   761512   0.00 │   8    0      nan │
│      coq-mathcomp-odd-order │  544.19   545.48  -0.24 │  2385522793549   2390573633185  -0.21 │  4083439209847   4083446233833  -0.00 │ 1224604  1224628  -0.00 │   0    0      nan │
│        coq-mathcomp-algebra │   59.94    60.06  -0.20 │   262541414944    262574077998  -0.01 │   360734917440    360745248834  -0.00 │  595748   595640   0.02 │   0    0      nan │
│                coq-compcert │  288.98   289.55  -0.20 │  1256128829469   1259949516667  -0.30 │  1889695186532   1889674199776   0.00 │ 1106768  1105288   0.13 │   0   16  -100.00 │
│    coq-metacoq-translations │   20.65    20.69  -0.19 │    89441613135     89577406541  -0.15 │   156616315386    156637039379  -0.01 │  965024   964724   0.03 │   0    0      nan │
│                    coq-core │  107.66   107.86  -0.19 │   386836723785    388507199752  -0.43 │   442628736797    442941677199  -0.07 │  283212   282012   0.43 │   0    0      nan │
│           coq-metacoq-pcuic │  552.78   553.77  -0.18 │  2414435829407   2419327315309  -0.20 │  3435300523738   3435295291695   0.00 │ 1572532  1572600  -0.00 │  17    0      nan │
│                 coq-coqutil │   36.14    36.20  -0.17 │   155318821751    156234922588  -0.59 │   210097103297    210065679560   0.01 │  606600   606420   0.03 │   4    4     0.00 │
│  coq-performance-tests-lite │  801.33   802.02  -0.09 │  3487717123896   3489514106560  -0.05 │  5944980916825   5945748923012  -0.01 │ 1824464  1824376   0.00 │   4    0      nan │
│               coq-fourcolor │ 1497.81  1499.08  -0.08 │  6553114668610   6558432577317  -0.08 │ 12071306775796  12071443113547  -0.00 │  790276   790100   0.02 │   0    0      nan │
│       coq-mathcomp-solvable │   82.67    82.72  -0.06 │   361716266127    362201217724  -0.13 │   555861366883    555859458482   0.00 │  665216   665176   0.01 │   0    0      nan │
│                 coq-unimath │ 4443.68  4446.31  -0.06 │ 19459725782109  19467413070501  -0.04 │ 38978842556263  38978147574345   0.00 │ 3258308  3258296   0.00 │   1    2   -50.00 │
│ coq-rewriter-perf-SuperFast │  959.47   959.88  -0.04 │  4193284706166   4194956445533  -0.04 │  7267386091616   7267277805883   0.00 │ 1146692  1146004   0.06 │   4    0      nan │
│                   coq-color │  215.72   215.78  -0.03 │   935925598435    934906735585   0.11 │  1327858412319   1327976690184  -0.01 │ 1149384  1149240   0.01 │   3   19   -84.21 │
│                  coq-flocq3 │   78.99    79.01  -0.03 │   344456810722    344438542506   0.01 │   457039338019    457015573020   0.01 │  988520   987480   0.11 │   8    0      nan │
│                coq-bedrock2 │  226.65   226.67  -0.01 │   989800336851    989147682440   0.07 │  1789929561135   1789800202338   0.01 │ 2052044  2051964   0.00 │   7    9   -22.22 │
│             coq-fiat-crypto │ 2900.40  2900.51  -0.00 │ 12654540663661  12654228561515   0.00 │ 23212322342439  23214502172361  -0.01 │ 3830868  3831432  -0.01 │  81   75     8.00 │
│                    coq-corn │  784.05   783.95   0.01 │  3422880815900   3422470316299   0.01 │  5281527133832   5281371855282   0.00 │  877876   879080  -0.14 │   0    0      nan │
│                     coq-vst │ 1095.44  1095.18   0.02 │  4783584355334   4782327568891   0.03 │  7716907947795   7716772227071   0.00 │ 2139244  2138988   0.01 │   2    1   100.00 │
│              coq-verdi-raft │  576.62   576.40   0.04 │  2517277590552   2516503440146   0.03 │  3928003006325   3928249815518  -0.01 │ 1261864  1261364   0.04 │   0    0      nan │
│     coq-metacoq-safechecker │  164.31   164.24   0.04 │   716223503733    715123487148   0.15 │   955417888886    955451925915  -0.00 │ 1187444  1187464  -0.00 │   2    2     0.00 │
│               coq-perennial │ 4105.54  4103.78   0.04 │ 17947923204944  17944445982201   0.02 │ 29797604609025  29797585061632   0.00 │ 2625596  2625724  -0.00 │   0    1  -100.00 │
│           coq-iris-examples │  426.42   425.83   0.14 │  1860337409209   1859100787139   0.07 │  2839873716368   2839954065822  -0.00 │ 1182988  1183176  -0.02 │  23    3   666.67 │
│        coq-metacoq-template │   79.04    78.91   0.16 │   338984634334    338508711607   0.14 │   510229517170    510212196536   0.00 │  911292   911504  -0.02 │  11    2   450.00 │
│                coq-rewriter │  386.81   386.17   0.17 │  1687417532632   1684884366798   0.15 │  2772559104656   2772471893041   0.00 │ 1065936  1063824   0.20 │   5    9   -44.44 │
│            coq-fiat-parsers │  345.98   345.38   0.17 │  1492046875130   1489125507936   0.20 │  2454112055930   2454127332207  -0.00 │ 2707704  2707520   0.01 │   1   24   -95.83 │
│                    coq-hott │  169.38   168.88   0.30 │   734411070241    734945075805  -0.07 │  1163086995431   1163127050170  -0.00 │  617948   617504   0.07 │   0    0      nan │
│       coq-mathcomp-fingroup │   22.27    22.16   0.50 │    97213753531     97065960104   0.15 │   140539599871    140534422136   0.00 │  486812   486952  -0.03 │   0    0      nan │
│      coq-mathcomp-ssreflect │   25.67    25.52   0.59 │   111582943744    111086441227   0.45 │   140134030223    140131410903   0.00 │  538812   537460   0.25 │   0    0      nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴───────────────────┘
```

</details>

File: tools/coqdep/coqdep.ml [hunk 4]
```diff
@@ -9,10 +9,16 @@
...
+module CD = Coqdeplib.Common
 
 let coqdep () =
-  init ();
+  let open CD in
+
+  (* Initialize coqdep, add files to dependency computation *)
+  if Array.length Sys.argv < 2 then CD.usage ();
+  let v_files = CD.init (List.tl (Array.to_list Sys.argv)) in
+  List.iter CD.treat_file_command_line v_files;
```
[the remark this answers is not in the document]
Comment (Alizter) [reply tools/coqdep/coqdep.ml]: Left over, cleaned up now.

File: tools/coqdep/common.ml (hunk 3 above)
Comment (ejgallego) [reply tools/coqdep/common.ml:533]: It means that at some point we should have `Coq_module.t` and `name : t \to string` etc...

What do you want me to do with this comment? [I put it as we were doing a hacking session]

Comment (SkySkimmer) [reply tools/coqdep/common.ml:533]: Is that different from renaming Dep_info to Coq_module?

Comment (gares) [reply tools/coqdep/common.ml:533]: Shouldn't this comment be removed now that there is a `Dep_info.t`?

**Review by gares — APPROVED** [review]

File: tools/coqdep/common.ml (hunk 3 above)
Comment (ejgallego) [reply tools/coqdep/common.ml:533]: What it means is that instead of `name : string` we should have `module : Coq_module.t` 

`Dep_info.t != Coq_module.t`

Comment (ejgallego) [reply tools/coqdep/common.ml:533]: I can take of it in the depending PR if you want, as you prefer.

Comment (SkySkimmer) [reply tools/coqdep/common.ml:533]: The comment is misleading then
```suggestion
  (* [name]] should be a [Coq_module.t] (which needs to be defined) *)
```

Comment (ejgallego) [reply tools/coqdep/common.ml]: ok fixed.

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**